src/Microsoft.VisualStudio.Threading/StackReentrantSemaphoreNestingViolationException.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Threading
+{
+    using System;
+    using System.Globalization;
+
+    /// <summary>
+    /// Exception which is thrown when the reentrancy contract of a <see cref="ReentrantSemaphore"/> created with <see cref="ReentrantSemaphore.ReentrancyMode.Stack"/> is violated.
+    /// </summary>
+    public class StackReentrantSemaphoreNestingViolationException : InvalidOperationException
```

Comment:
I'm thinking a bit more now how this is a whole exception type for a *very* specific error condition. What if we call it `IllegalSemaphoreUsageException` and change its public constructor to take the standard `string, Exception` parameters? I think that will still accomplish your goals, but it might be useful for more scenarios in the future so we don't have to keep introducing new public API as error conditions may be discovered.

Reply:
That is a bit of a philosophical question, but I tend to error on the side of specific errors for specific cases.  I believe in this case, using a more generic exception would be fine as I don't see another way to violate usage of the semaphore in this execution path.  However, it is essential to keep the first fail (that puts things into the failed mode) separate from the subsequent failures (semaphore already broken/faulted).

Reply:
Updated to IllegalSemaphoreUsageException

Reply:
It has been a bit since I have done a github PR.  I think it needs a re-review after latest changes.  Thanks @AArnott!